### PR TITLE
Code review

### PR DESCRIPTION
**Pull request by knstvk**

### CODE REVIEW

File: modules/core/src/com/haulmont/addon/ldap/core/service/UserSynchronizationServiceBean.java [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+
+        List<CommonMatchingRule> result = rulesToApply.stream().filter(r -> !CUSTOM.equals(r.getRuleType())).collect(Collectors.toList());
+        List<CustomLdapMatchingRuleWrapper> customRules = matchingRuleDao.getCustomMatchingRules();
+        rulesToApply.stream()
+                .filter(r -> CUSTOM.equals(r.getRuleType()))
+                .forEach(customRuleDto -> {
+                    CustomLdapMatchingRuleWrapper wrapper = customRules.stream()
+                            .filter(cr -> cr.getMatchingRuleId().equals(customRuleDto.getMatchingRuleId()))
+                            .findFirst()
+                            .get();
```
Comment (knstvk) [inline modules/core/src/com/haulmont/addon/ldap/core/service/UserSynchronizationServiceBean.java]: Consider using `orElseThrow` with a specific message

File: modules/core/src/com/haulmont/addon/ldap/core/service/UserSynchronizationServiceBean.java [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    private CubaUserDao cubaUserDao;
+
+    @Inject
+    private MatchingRuleDao matchingRuleDao;
+
+    @Inject
+    private MatchingRuleApplier matchingRuleApplier;
+
+    @Inject
+    private ApplicationEventPublisher applicationEventPublisher;
```
Comment (knstvk) [inline modules/core/src/com/haulmont/addon/ldap/core/service/UserSynchronizationServiceBean.java]: Use CUBA `Events` bean: https://doc.cuba-platform.com/manual-6.8/events.html

File: modules/core/src/com/haulmont/addon/ldap/core/spring/events/AfterUserUpdatedFromLdapEvent.java [hunk 3]
```diff
@@ -0,0 +1,17 @@
+package com.haulmont.addon.ldap.core.spring.events;
+
+import com.haulmont.addon.ldap.core.rule.LdapMatchingRuleContext;
+import com.haulmont.cuba.security.entity.User;
+import org.springframework.context.ApplicationEvent;
+
+public class AfterUserUpdatedFromLdapEvent extends ApplicationEvent {
+
+    private final LdapMatchingRuleContext ldapMatchingRuleContext;
```
Comment (knstvk) [inline modules/core/src/com/haulmont/addon/ldap/core/spring/events/AfterUserUpdatedFromLdapEvent.java]: Here and other events: add getters?

File: modules/web/src/com/haulmont/addon/ldap/web/ldapconfig/LdapConfigEdit.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+import static com.haulmont.cuba.gui.components.Frame.NotificationType.HUMANIZED;
+import static com.haulmont.cuba.gui.components.Frame.NotificationType.WARNING;
+
+public class LdapConfigEdit extends AbstractEditor<LdapConfig> {
+
+    @Inject
+    private LdapService ldapService;
+
+    @Named("ldapUserAttributesDs")
```
Comment (knstvk) [inline modules/web/src/com/haulmont/addon/ldap/web/ldapconfig/LdapConfigEdit.java]: `@Inject` if the field name is the same as component id

File: modules/web/src/com/haulmont/addon/ldap/web/screens/MatchingRuleScreen.java [hunk 5]
```diff
@@ -0,0 +1,450 @@
...
+        EditAction customEdit = new EditAction(matchingRuleTable) {
+            @Override
+            protected void internalOpenEditor(CollectionDatasource datasource, Entity existingItem, Datasource parentDs, Map<String, Object> params) {
+                super.internalOpenEditor(datasource, existingItem, datasource, params);
+            }
+
+            @Override
+            public String getWindowId() {
+                AbstractCommonMatchingRule rule = matchingRuleTable.getSingleSelected();
+                if (DEFAULT.equals(rule.getRuleType())) {
```
Comment (knstvk) [inline modules/web/src/com/haulmont/addon/ldap/web/screens/MatchingRuleScreen.java]: switch

File: modules/web/src/com/haulmont/addon/ldap/web/screens/MatchingRuleScreen.java [hunk 6]
```diff
@@ -0,0 +1,450 @@
...
+
+
+            }
+        };
+
+        customEdit.setBeforeActionPerformedHandler(customEditBeforeActionPerformedHandler);
+        customEdit.setAfterCommitHandler(customAfterCommitHandler);
+
+
+        RemoveAction.BeforeActionPerformedHandler customRemoveBeforeActionPerformedHandler = new RemoveAction.BeforeActionPerformedHandler() {
```
Comment (knstvk) [inline modules/web/src/com/haulmont/addon/ldap/web/screens/MatchingRuleScreen.java]: Replace with lambda

File: modules/web/src/com/haulmont/addon/ldap/web/screens/messages.properties [hunk 7]
```diff
@@ -0,0 +1,73 @@
+matchingRuleScreenCaption = LDAP matching rule
+ldapSettingsScreenCaption = LDAP config
+matchingRuleTableConditionColumn = Condition
+matchingRuleTableCubaColumn = Cuba
```
Comment (knstvk) [inline modules/web/src/com/haulmont/addon/ldap/web/screens/messages.properties]: CUBA is always in capital letters

File: modules/web/src/com/haulmont/addon/ldap/web/screens/messages.properties [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+testMatchingRuleScreen = Test matching Rule
+testRuleScreenTestButton = Test rules
+testRuleScreenRuleTable = Applied matching rules
+testRuleScreenRoleTable = Applied roles
+testMatchingRuleTableTypeColumn = Type
+testMatchingRuleTableOptionsColumn = Options
+testMatchingRuleTableResultColumn = Cuba
+testMatchingRuleTableConditionColumn = Condition
+testRuleScreenUserName = User login
+settingsScreenUrl = Url
```
Comment (knstvk) [inline modules/web/src/com/haulmont/addon/ldap/web/screens/messages.properties]: All abbreviations in capital letters: URL

File: modules/web/src/com/haulmont/addon/ldap/web/screens/messages.properties [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+settingsScreenAccessGroupAttr = Access group attribute name
+settingsScreenAccountExpAttr = Account expires attribute name
+matchingRuleScreenCommitButton = Apply
+matchingRuleScreenCancelButton = Rollback changes
+matchingRuleScreenCommitDialogTitle = Apply
+matchingRuleScreenCommitDialogMsg = Apply?
+matchingRuleScreenCancelDialogTitle = Rollback changes
+matchingRuleScreenCancelDialogMsg = Rollback changes?
+testRuleScreenAppliedGroup = Group
+testRuleScreenUserNotInLdap = Such user don't exists in LDAP
```
Comment (knstvk) [inline modules/web/src/com/haulmont/addon/ldap/web/screens/messages.properties]: doesn't

File: modules/web/src/com/haulmont/addon/ldap/web/screens/messages.properties [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+testRuleScreenUserNotInLdapCaption = Warning
+matchingRuleTableOrderColumn = Order
+matchingRuleTableDefaultRuleOrder = LAST
+matchingRuleScreenEmptyOrderCaption = Empty order
+matchingRuleScreenEmptyOrder = Some matching rules has empty order
+matchingRuleScreenDuplicationOrderCaption = Duplication order
```
Comment (knstvk) [inline modules/web/src/com/haulmont/addon/ldap/web/screens/messages.properties]: Duplicate?

File: modules/web/src/com/haulmont/addon/ldap/web/screens/matchingRuleScreen.xml [hunk 11]
```diff
@@ -0,0 +1,150 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
Comment (knstvk) [inline modules/web/src/com/haulmont/addon/ldap/web/screens/matchingRuleScreen.xml]: matching-rule-screen.xml

File: modules/web/src/com/haulmont/addon/ldap/web/scriptingmatchingrule/EmptyGroovyScriptHelpAction.java [hunk 12]
```diff
@@ -0,0 +1,121 @@
+package com.haulmont.addon.ldap.web.scriptingmatchingrule;
+
+import com.haulmont.cuba.gui.components.Action;
+import com.haulmont.cuba.gui.components.Component;
+import com.haulmont.cuba.gui.components.KeyCombination;
+
+import java.beans.PropertyChangeListener;
+import java.util.Collection;
+
+public class EmptyGroovyScriptHelpAction implements Action{
```
Comment (knstvk) [inline modules/web/src/com/haulmont/addon/ldap/web/scriptingmatchingrule/EmptyGroovyScriptHelpAction.java]: `extends AbstractAction` or `BaseAction`?

File: modules/global/src/com/haulmont/addon/ldap/entity/MatchingRuleType.java [hunk 13]
```diff
@@ -0,0 +1,25 @@
+package com.haulmont.addon.ldap.entity;
+
+public enum MatchingRuleType {
```
[the remark this answers is not in the document]
Comment (antmarr) [reply modules/global/src/com/haulmont/addon/ldap/entity/MatchingRuleType.java]: Accepted. Will do using standard CUBA enums.

File: modules/global/src/com/haulmont/addon/ldap/entity/AbstractCommonMatchingRule.java [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+
+import javax.persistence.*;
+import javax.validation.constraints.NotNull;
+
+@MappedSuperclass
+public abstract class AbstractCommonMatchingRule extends StandardEntity implements CommonMatchingRule {
+    private static final long serialVersionUID = 1956446424046023194L;
+
+    @Column(name = "RULE_TYPE")
+    @Enumerated(EnumType.STRING)
```
[the remark this answers is not in the document]
Comment (antmarr) [reply modules/global/src/com/haulmont/addon/ldap/entity/AbstractCommonMatchingRule.java]: Accepted. Will do using standard CUBA enums.

File: modules/global/src/com/haulmont/addon/ldap/entity/AbstractCommonMatchingRule.java [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+    private MatchingRuleType ruleType;
+
+    @Composition
+    @OnDelete(DeletePolicy.CASCADE)
+    @OneToOne(fetch = FetchType.LAZY, optional = false, cascade = CascadeType.ALL)
```
[the remark this answers is not in the document]
Comment (antmarr) [reply modules/global/src/com/haulmont/addon/ldap/entity/AbstractCommonMatchingRule.java]: Is it requirement or recommendation?
In my case I didn't see any problems with JPA cascade.

File: modules/global/src/com/haulmont/addon/ldap/entity/AbstractCommonMatchingRule.java [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+    @JoinColumn(name = "MATCHING_RULE_STATUS_ID")
+    @NotNull
+    private MatchingRuleStatus status = new MatchingRuleStatus();
```
[the remark this answers is not in the document]
Comment (antmarr) [reply modules/global/src/com/haulmont/addon/ldap/entity/AbstractCommonMatchingRule.java]: Accepted. Will try to initialize these fields in screen's callbacks.

File: modules/global/src/com/haulmont/addon/ldap/entity/SimpleMatchingRule.java [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+
+@DiscriminatorValue("SIMPLE")
+@Entity(name = "ldap$SimpleMatchingRule")
+public class SimpleMatchingRule extends AbstractDbStoredMatchingRule {
+    private static final long serialVersionUID = -2383286286785487816L;
+
+
+    @Composition
+    @OnDelete(DeletePolicy.CASCADE)
+    @OneToMany(mappedBy = "simpleMatchingRule", cascade = CascadeType.ALL, orphanRemoval = true)
```
[the remark this answers is not in the document]
Comment (antmarr) [reply modules/global/src/com/haulmont/addon/ldap/entity/SimpleMatchingRule.java]: Is it requirement or recommendation?
In my case I didn't see any problems with JPA cascade.

File: modules/global/src/com/haulmont/addon/ldap/dto/LdapContextDto.java [hunk 18]
```diff
@@ -0,0 +1,29 @@
+package com.haulmont.addon.ldap.dto;
+
+
+import java.io.Serializable;
+
+public class LdapContextDto implements Serializable{
```
[the remark this answers is not in the document]
Comment (antmarr) [reply modules/global/src/com/haulmont/addon/ldap/dto/LdapContextDto.java]: Ok.

File: modules/global/src/com/haulmont/addon/ldap/dto/TestUserSynchronizationDto.java [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+
+import com.haulmont.addon.ldap.entity.AbstractCommonMatchingRule;
+import com.haulmont.cuba.security.entity.Group;
+import com.haulmont.cuba.security.entity.Role;
+
+import java.io.Serializable;
+import java.util.LinkedHashSet;
+import java.util.Set;
+
+public class TestUserSynchronizationDto implements Serializable{
```
[the remark this answers is not in the document]
Comment (antmarr) [reply modules/global/src/com/haulmont/addon/ldap/dto/TestUserSynchronizationDto.java]: Ok.

File: modules/global/src/com/haulmont/addon/ldap/utils/MatchingRuleUtils.java [hunk 20]
```diff
@@ -0,0 +1,111 @@
...
+                sb.append("=");
+                sb.append(simpleRuleCondition.getAttributeValue());
+                sb.append("\n");
+            }
+        }
+        return sb.toString();
+    }
+
+    public String generateMatchingRuleOptionsColumn(AbstractCommonMatchingRule entity) {
+        if (CUSTOM.equals(entity.getRuleType())) return StringUtils.EMPTY;
```
[the remark this answers is not in the document]
Comment (antmarr) [reply modules/global/src/com/haulmont/addon/ldap/utils/MatchingRuleUtils.java]: Ok.

File: modules/global/src/com/haulmont/addon/ldap/utils/MatchingRuleUtils.java [hunk 21]
```diff
@@ -0,0 +1,111 @@
...
+        }
+        sb.append("\n");
+        sb.append("Access group: ");
+        sb.append(dbStoredMatchingRule.getAccessGroup() == null ? StringUtils.EMPTY : dbStoredMatchingRule.getAccessGroup().getName());
+
+        return sb.toString();
+    }
+
+    public String generateMatchingRuleTableConditionColumn(AbstractCommonMatchingRule entity) {
+        if (SIMPLE.equals(entity.getRuleType())) {
```
[the remark this answers is not in the document]
Comment (antmarr) [reply modules/global/src/com/haulmont/addon/ldap/utils/MatchingRuleUtils.java]: Ok.

File: build.gradle [hunk 22]
```diff
@@ -0,0 +1,243 @@
+
+buildscript {
+    ext.cubaVersion = '6.7.4'
```
[the remark this answers is not in the document]
Comment (antmarr) [reply build.gradle]: Ok.

File: build.gradle [hunk 23]
```diff
@@ -0,0 +1,243 @@
...
+
+cuba {
+    artifact {
+        group = 'com.haulmont.addon.ldap'
+        version = '0.1'
+        isSnapshot = true
+    }
+    tomcat {
+        dir = "$project.rootDir/deploy/tomcat"
+        port = 8081
```
[the remark this answers is not in the document]
Comment (antmarr) [reply build.gradle]: Ok.

File: build.gradle [hunk 24]
```diff
@@ -0,0 +1,243 @@
...
+
+
+apply(plugin: 'idea')
+apply(plugin: 'cuba')
+
+cuba {
+    artifact {
+        group = 'com.haulmont.addon.ldap'
+        version = '0.1'
+        isSnapshot = true
```
[the remark this answers is not in the document]
Comment (antmarr) [reply build.gradle]: Do you mean isSnapshot = false?

File: modules/core/src/com/haulmont/addon/ldap/core/dao/CubaUserDao.java [hunk 25]
```diff
@@ -0,0 +1,69 @@
...
+import org.springframework.transaction.annotation.Transactional;
+
+import javax.inject.Inject;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static com.haulmont.addon.ldap.core.dao.CubaUserDao.NAME;
+
+@Service(NAME)
```
[the remark this answers is not in the document]
Comment (antmarr) [reply modules/core/src/com/haulmont/addon/ldap/core/dao/CubaUserDao.java]: Ok.

File: modules/core/src/com/haulmont/addon/ldap/core/dao/CubaUserDao.java [hunk 26]
```diff
@@ -0,0 +1,69 @@
...
+        TypedQuery<User> query = persistence.getEntityManager()
+                .createQuery("select distinct cu from sec$User cu where cu.login = :login", User.class);
+        query.setViewName("sec-user-view-with-group-roles");
+        return query.getResultList();
+    }
+
+    @Transactional
+    public void saveCubaUser(User cubaUser, User beforeRulesApplyUserState, LdapMatchingRuleContext ldapMatchingRuleContext) {
+        EntityManager entityManager = persistence.getEntityManager();
+        User mergedUser = PersistenceHelper.isNew(cubaUser) ? cubaUser : entityManager.merge(cubaUser);
```
[the remark this answers is not in the document]
Comment (antmarr) [reply modules/core/src/com/haulmont/addon/ldap/core/dao/CubaUserDao.java]: Accepted.

File: modules/core/src/com/haulmont/addon/ldap/core/dao/LdapUserAttributeDao.java [hunk 27]
```diff
@@ -0,0 +1,54 @@
...
+import com.haulmont.cuba.core.global.PersistenceHelper;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import javax.inject.Inject;
+import java.util.List;
+
+import static com.haulmont.addon.ldap.core.dao.LdapUserAttributeDao.NAME;
+
+@Service(NAME)
```
[the remark this answers is not in the document]
Comment (antmarr) [reply modules/core/src/com/haulmont/addon/ldap/core/dao/LdapUserAttributeDao.java]: Ok.

File: modules/core/src/com/haulmont/addon/ldap/core/dao/LdapUserDao.java [hunk 28]
```diff
@@ -0,0 +1,146 @@
...
+import org.springframework.ldap.support.LdapUtils;
+import org.springframework.stereotype.Service;
+
+import javax.inject.Inject;
+import java.util.List;
+import java.util.Locale;
+
+import static com.haulmont.addon.ldap.core.dao.LdapUserDao.NAME;
+
+@Service(NAME)
```
[the remark this answers is not in the document]
Comment (antmarr) [reply modules/core/src/com/haulmont/addon/ldap/core/dao/LdapUserDao.java]: Ok.

File: modules/core/src/com/haulmont/addon/ldap/core/dao/MatchingRuleDao.java [hunk 29]
```diff
@@ -0,0 +1,148 @@
...
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static com.haulmont.addon.ldap.core.dao.MatchingRuleDao.NAME;
+import static com.haulmont.addon.ldap.entity.MatchingRuleType.DEFAULT;
+
+@Service(NAME)
```
[the remark this answers is not in the document]
Comment (antmarr) [reply modules/core/src/com/haulmont/addon/ldap/core/dao/MatchingRuleDao.java]: Ok.

File: modules/core/src/com/haulmont/addon/ldap/core/messages.properties [hunk 30]
```diff
@@ -0,0 +1,13 @@
+onlySingleDefaultRule=Only single default rule can present
```
[the remark this answers is not in the document]
Comment (antmarr) [reply modules/core/src/com/haulmont/addon/ldap/core/messages.properties]: Ok.

File: modules/core/src/com/haulmont/addon/ldap/core/messages.properties [hunk 31]
```diff
@@ -0,0 +1,13 @@
...
+LoginException.InvalidLoginOrPassword=Unknown login name or bad password: %s
+successLdapLogin=User successfully login in LDAP: %s
```
[the remark this answers is not in the document]
Comment (antmarr) [reply modules/core/src/com/haulmont/addon/ldap/core/messages.properties]: Ok

File: modules/core/src/com/haulmont/addon/ldap/core/messages.properties [hunk 32]
```diff
@@ -0,0 +1,13 @@
...
+errorDuringLdapLogin=Error during LDAP login: %s
+userSyncStart=Start synchronization for user: %s
+userSyncEnd=End synchronization for user: %s
+errorDuringLdapSync=Error during LDAP synchronization: %s
+userCreatedFromLdap=User created from LDAP: %s
+userDeactivatedFromLdap=User deactivated from LDAP: %s
+userGetCommonInfoFromLdap=User get common info from LDAP: %s
```
[the remark this answers is not in the document]
Comment (antmarr) [reply modules/core/src/com/haulmont/addon/ldap/core/messages.properties]: Ok

File: modules/core/src/com/haulmont/addon/ldap/core/messages.properties [hunk 33]
```diff
@@ -0,0 +1,13 @@
...
+errorDuringGroovyScriptEvaluation=Error during script evaluation for user: %s
+testGroovyScriptResultNonBoolean = Groovy script return non boolean value: %s
```
[the remark this answers is not in the document]
Comment (antmarr) [reply modules/core/src/com/haulmont/addon/ldap/core/messages.properties]: Ok.

File: modules/core/src/com/haulmont/addon/ldap/core/messages.properties [hunk 34]
```diff
@@ -0,0 +1,13 @@
...
+userMustExistsInLdap = User with login %s must exists in LDAP
```
[the remark this answers is not in the document]
Comment (antmarr) [reply modules/core/src/com/haulmont/addon/ldap/core/messages.properties]: Ok

File: modules/core/src/com/haulmont/addon/ldap/core/dao/MatchingRuleDao.java [hunk 35]
```diff
@@ -0,0 +1,148 @@
...
+
+    @Transactional
+    public void saveMatchingRules(List<AbstractCommonMatchingRule> matchingRules, List<AbstractCommonMatchingRule> matchingRulesToDelete) {
+
+        List<CommonMatchingRule> defaultRules = matchingRules.stream().filter(mr -> DEFAULT.equals(mr.getRuleType())).collect(Collectors.toList());
+        if (defaultRules.size() != 1) {
+            throw new RuntimeException(messages.formatMessage(MatchingRuleDao.class, "onlySingleDefaultRule"));
+        }
+        EntityManager entityManager = persistence.getEntityManager();
+        matchingRules.forEach(mr -> {
```
[the remark this answers is not in the document]
Comment (antmarr) [reply modules/core/src/com/haulmont/addon/ldap/core/dao/MatchingRuleDao.java]: Ok.

File: modules/core/src/com/haulmont/addon/ldap/core/dao/MatchingRuleDao.java [hunk 36]
```diff
@@ -0,0 +1,148 @@
...
+            if (MatchingRuleType.CUSTOM.equals(mr.getRuleType())) {
```
[the remark this answers is not in the document]
Comment (antmarr) [reply modules/core/src/com/haulmont/addon/ldap/core/dao/MatchingRuleDao.java]: Ok.

File: modules/global/src/com/haulmont/addon/ldap/entity/UserSynchronizationResultEnum.java [hunk 37]
```diff
@@ -0,0 +1,32 @@
+package com.haulmont.addon.ldap.entity;
+
+import com.haulmont.chile.core.datatypes.impl.EnumClass;
+
+import javax.annotation.Nullable;
+
+
+public enum UserSynchronizationResultEnum implements EnumClass<String> {
+    SuccessSync("SUCCESS_SYNC"),
```
[the remark this answers is not in the document]
Comment (antmarr) [reply modules/global/src/com/haulmont/addon/ldap/entity/UserSynchronizationResultEnum.java]: Ok.

File: modules/global/src/com/haulmont/addon/ldap/entity/MatchingRuleType.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (antmarr) [reply modules/global/src/com/haulmont/addon/ldap/entity/MatchingRuleType.java]: Done.

File: modules/global/src/com/haulmont/addon/ldap/entity/AbstractCommonMatchingRule.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (antmarr) [reply modules/global/src/com/haulmont/addon/ldap/entity/AbstractCommonMatchingRule.java]: Done.

File: modules/global/src/com/haulmont/addon/ldap/entity/AbstractCommonMatchingRule.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (antmarr) [reply modules/global/src/com/haulmont/addon/ldap/entity/AbstractCommonMatchingRule.java]: Done.

File: modules/global/src/com/haulmont/addon/ldap/entity/UserSynchronizationResultEnum.java (hunk 37 above)
[the remark this answers is not in the document]
Comment (antmarr) [reply modules/global/src/com/haulmont/addon/ldap/entity/UserSynchronizationResultEnum.java]: Done.

File: modules/global/src/com/haulmont/addon/ldap/dto/LdapContextDto.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (antmarr) [reply modules/global/src/com/haulmont/addon/ldap/dto/LdapContextDto.java]: Done.

File: modules/global/src/com/haulmont/addon/ldap/dto/TestUserSynchronizationDto.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (antmarr) [reply modules/global/src/com/haulmont/addon/ldap/dto/TestUserSynchronizationDto.java]: Done.

File: build.gradle (hunk 23 above)
[the remark this answers is not in the document]
Comment (antmarr) [reply build.gradle]: Done. Change to 8080

File: modules/core/src/com/haulmont/addon/ldap/core/dao/CubaUserDao.java (hunk 25 above)
[the remark this answers is not in the document]
Comment (antmarr) [reply modules/core/src/com/haulmont/addon/ldap/core/dao/CubaUserDao.java]: Done.

File: modules/core/src/com/haulmont/addon/ldap/core/dao/LdapUserAttributeDao.java (hunk 27 above)
[the remark this answers is not in the document]
Comment (antmarr) [reply modules/core/src/com/haulmont/addon/ldap/core/dao/LdapUserAttributeDao.java]: Done.

File: modules/core/src/com/haulmont/addon/ldap/core/dao/LdapUserDao.java (hunk 28 above)
[the remark this answers is not in the document]
Comment (antmarr) [reply modules/core/src/com/haulmont/addon/ldap/core/dao/LdapUserDao.java]: Done.

File: modules/core/src/com/haulmont/addon/ldap/core/dao/MatchingRuleDao.java (hunk 29 above)
[the remark this answers is not in the document]
Comment (antmarr) [reply modules/core/src/com/haulmont/addon/ldap/core/dao/MatchingRuleDao.java]: Done.

File: modules/core/src/com/haulmont/addon/ldap/core/messages.properties (hunk 30 above)
[the remark this answers is not in the document]
Comment (antmarr) [reply modules/core/src/com/haulmont/addon/ldap/core/messages.properties]: Done.

File: modules/core/src/com/haulmont/addon/ldap/core/messages.properties (hunk 31 above)
[the remark this answers is not in the document]
Comment (antmarr) [reply modules/core/src/com/haulmont/addon/ldap/core/messages.properties]: Done.

File: modules/core/src/com/haulmont/addon/ldap/core/messages.properties (hunk 32 above)
[the remark this answers is not in the document]
Comment (antmarr) [reply modules/core/src/com/haulmont/addon/ldap/core/messages.properties]: Done.

File: modules/core/src/com/haulmont/addon/ldap/core/messages.properties (hunk 33 above)
[the remark this answers is not in the document]
Comment (antmarr) [reply modules/core/src/com/haulmont/addon/ldap/core/messages.properties]: Done.

File: modules/core/src/com/haulmont/addon/ldap/core/messages.properties (hunk 34 above)
[the remark this answers is not in the document]
Comment (antmarr) [reply modules/core/src/com/haulmont/addon/ldap/core/messages.properties]: Done

File: modules/core/src/com/haulmont/addon/ldap/core/dao/MatchingRuleDao.java (hunk 35 above)
[the remark this answers is not in the document]
Comment (antmarr) [reply modules/core/src/com/haulmont/addon/ldap/core/dao/MatchingRuleDao.java]: Done.

File: modules/core/src/com/haulmont/addon/ldap/core/dao/MatchingRuleDao.java (hunk 36 above)
[the remark this answers is not in the document]
Comment (antmarr) [reply modules/core/src/com/haulmont/addon/ldap/core/dao/MatchingRuleDao.java]: Done.

File: modules/core/src/com/haulmont/addon/ldap/core/dao/UserSynchronizationLogDao.java [hunk 38]
```diff
@@ -0,0 +1,213 @@
...
+import static java.lang.Boolean.FALSE;
+import static java.lang.Boolean.TRUE;
+import static java.util.stream.Collectors.toList;
+
+@Service(NAME)
+public class UserSynchronizationLogDao {
+
+    public final static String NAME = "ldap_UserSynchronizationLogDao";
+
+    private final String INDENT = "   ";
```
[the remark this answers is not in the document]
Comment (antmarr) [reply modules/core/src/com/haulmont/addon/ldap/core/dao/UserSynchronizationLogDao.java]: Add static

File: modules/core/src/com/haulmont/addon/ldap/core/dao/UserSynchronizationLogDao.java [hunk 39]
```diff
@@ -0,0 +1,213 @@
...
+    }
+
+    private String getLdapAttributes(Map<String, Object> ldapAttributes) {
+        StringBuilder sb = new StringBuilder();
+        for (Map.Entry<String, Object> me : ldapAttributes.entrySet()) {
+            String attrName = me.getKey();
+            sb.append(attrName);
+            sb.append(":");
+            if (me.getValue() instanceof List) {
+                List<String> list = ((List<Object>) me.getValue()).stream().map(v -> v == null ? "null" : v.toString()).collect(toList());
```
[the remark this answers is not in the document]
Comment (antmarr) [reply modules/core/src/com/haulmont/addon/ldap/core/dao/UserSynchronizationLogDao.java]: It is one line stream.

File: modules/core/src/com/haulmont/addon/ldap/core/service/UserSynchronizationServiceBean.java (hunk 1 above)
Comment (antmarr) [reply modules/core/src/com/haulmont/addon/ldap/core/service/UserSynchronizationServiceBean.java]: Done.

File: modules/web/src/com/haulmont/addon/ldap/web/screens/messages.properties (hunk 7 above)
Comment (antmarr) [reply modules/web/src/com/haulmont/addon/ldap/web/screens/messages.properties]: Done.

File: modules/web/src/com/haulmont/addon/ldap/web/screens/messages.properties (hunk 8 above)
Comment (antmarr) [reply modules/web/src/com/haulmont/addon/ldap/web/screens/messages.properties]: Done

File: modules/web/src/com/haulmont/addon/ldap/web/screens/messages.properties (hunk 9 above)
Comment (antmarr) [reply modules/web/src/com/haulmont/addon/ldap/web/screens/messages.properties]: Done.

File: modules/web/src/com/haulmont/addon/ldap/web/screens/messages.properties (hunk 10 above)
Comment (antmarr) [reply modules/web/src/com/haulmont/addon/ldap/web/screens/messages.properties]: Done.

File: modules/web/src/com/haulmont/addon/ldap/web/screens/matchingRuleScreen.xml (hunk 11 above)
Comment (antmarr) [reply modules/web/src/com/haulmont/addon/ldap/web/screens/matchingRuleScreen.xml]: Done.

File: modules/web/src/com/haulmont/addon/ldap/web/scriptingmatchingrule/EmptyGroovyScriptHelpAction.java (hunk 12 above)
Comment (antmarr) [reply modules/web/src/com/haulmont/addon/ldap/web/scriptingmatchingrule/EmptyGroovyScriptHelpAction.java]: Done.

File: modules/global/src/com/haulmont/addon/ldap/entity/AbstractCommonMatchingRule.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (antmarr) [reply modules/global/src/com/haulmont/addon/ldap/entity/AbstractCommonMatchingRule.java]: Suspended. Waiting for details.

File: modules/core/src/com/haulmont/addon/ldap/core/dao/CubaUserDao.java (hunk 26 above)
[the remark this answers is not in the document]
Comment (antmarr) [reply modules/core/src/com/haulmont/addon/ldap/core/dao/CubaUserDao.java]: Done.

File: modules/global/src/com/haulmont/addon/ldap/entity/SimpleMatchingRule.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (knstvk) [reply modules/global/src/com/haulmont/addon/ldap/entity/SimpleMatchingRule.java]: See details here: https://doc.cuba-platform.com/manual-6.8/entity_attr_annotations.html#oneToMany_annotation
EntityLog also won't work.

**Pull request closed without merging**